packages/snaps-controllers/src/snaps/SnapController.ts

```
@@ -1713,30 +1734,100 @@ export class SnapController extends BaseController<
+        });
+
+        throw error;
+      }
 
-      await this.authorize(origin, snapId);
+      await this.authorize(snapId, permissionsApproval);
 
-      await this.#startSnap({
+      const resultApproval = this.#createApproval({
+        origin,
         snapId,
-        sourceCode,
+        type: SNAP_APPROVAL_RESULT,
       });
 
+      try {
+        const { sourceCode } = this.#getRuntimeExpect(snapId);
+        if (sourceCode) {
+          await this.#startSnap({
+            snapId,
+            sourceCode,
+          });
+        } else {
+          throw Error('no source code found');
+        }
+      } catch (error) {
+        this.#updateApproval(resultApproval.id, {
+          loading: false,
+          error: SnapInstallError.StartFailed,
+        });
+
+        throw error;
+      }
+
       const truncated = this.getTruncatedExpect(snapId);
 
+      this.#updateApproval(resultApproval.id, { loading: false });
+
       this.messagingSystem.publish(`SnapController:snapInstalled`, truncated);
+
       return truncated;
     } catch (error) {
       logError(`Error when adding snap.`, error);
-
       throw error;
     }
   }
 
+  #createApproval({
+    origin,
+    snapId,
+    type,
+  }: {
+    origin: string;
+    snapId: ValidatedSnapId;
+    type: string;
+  }): PendingApproval {
+    const id = nanoid();
+    const promise = this.messagingSystem.call(
+      'ApprovalController:addRequest',
+      {
+        origin,
+        id,
+        type,
+        requestData: {
+          // Mirror previous installation metadata
+          metadata: { id, origin: snapId, dappOrigin: origin },
+          snapId,
+        },
+        requestState: {
+          loading: true,
+          error: null,
+        },
+      },
+      true,
+    );
+
+    return { id, promise };
+  }
+
+  #updateApproval(id: string, requestState: Record<string, Json>) {
+    this.messagingSystem.call('ApprovalController:updateRequestState', {
```

This may throw. Do you think we should catch and ignore errors here? 🤔